### PR TITLE
Use cloud-init tool to setup an Instance

### PR DESCRIPTION
**Pull request by nesitor**
Problem: Setup an instance using a cloud OS image file was impossible without a custom init file

Solution: Use the cloud-init tool already integrated in some cloud distributions to setup the network and system configuration.

### CODE REVIEW

File: kernels/microvm-kernel-x86_64-5.10.config [hunk 1]
```diff
@@ -2156,7 +2156,7 @@ CONFIG_OVERLAY_FS_REDIRECT_ALWAYS_FOLLOW=y
 #
 # CD-ROM/DVD Filesystems
 #
-# CONFIG_ISO9660_FS is not set
+CONFIG_ISO9660_FS=y
```
Comment (hoh) [inline kernels/microvm-kernel-x86_64-5.10.config:2159]: Can you confirm that this is the only change required in the config file of the kernel to work ?

File: vm_supervisor/vm/firecracker/instance.py [hunk 2]
```diff
@@ -50,14 +52,14 @@ class AlephFirecrackerInstance(AlephFirecrackerExecutable):
...
-        hardware_resources: MachineResources = MachineResources(),
-        tap_interface: Optional[TapInterface] = None,
+            self,
+            vm_id: int,
+            vm_hash: ItemHash,
+            resources: AlephInstanceResources,
+            enable_networking: bool = False,
+            enable_console: Optional[bool] = None,
+            hardware_resources: MachineResources = MachineResources(),
+            tap_interface: Optional[TapInterface] = None,
```
Comment (hoh) [inline vm_supervisor/vm/firecracker/instance.py:62]: Please make sure that the PR passes the code quality tests: https://github.com/aleph-im/aleph-vm/actions/runs/5323885887/jobs/9642375702

`black`, `isort` and `mypy` all report errors on these changes.

I know that our global test suite is broken due to issues with the testing on virtual servers, but the rest of the tests should pass.

File: vm_supervisor/vm/firecracker/instance.py [hunk 3]
```diff
@@ -115,5 +120,89 @@ async def wait_for_init(self) -> None:
...
         """Configure the VM by sending configuration info to it's init"""
-        # TODO: Implement Cloud-init interface
         pass
+
+    async def _create_user_data_file(self) -> str:
+        """Creates user data configuration file for cloud-init tool"""
+
+        ssh_authorized_keys = self.resources.message_content.authorized_keys \
+            if self.resources.message_content.authorized_keys \
+            else []
```
Comment (hoh) [inline vm_supervisor/vm/firecracker/instance.py:130]:
```suggestion
        ssh_authorized_keys = self.resources.message_content.authorized_keys or []
```

File: vm_supervisor/vm/firecracker/instance.py [hunk 4]
```diff
@@ -115,5 +120,89 @@ async def wait_for_init(self) -> None:
...
+        return network_yaml_file.name
+
+    async def _create_cloud_init_drive(self) -> Drive:
+        """Creates the cloud-init volume to configure and setup the VM"""
+
+        assert (
+                self.enable_networking and self.tap_interface
+        ), f"Network not enabled for VM {self.vm_id}"
+
+        disk_image_path = Path(f"/opt/cloud-init-{self.vm_hash}.img")
```
Comment (hoh) [inline vm_supervisor/vm/firecracker/instance.py:192]: All data on disk must be stored in either `/var/lib/aleph` or `/var/cache/aleph`.

File: vm_supervisor/vm/firecracker/instance.py [hunk 5]
```diff
@@ -115,5 +120,89 @@ async def wait_for_init(self) -> None:
...
+        """Creates user data configuration file for cloud-init tool"""
+
+        ssh_authorized_keys = self.resources.message_content.authorized_keys \
+            if self.resources.message_content.authorized_keys \
+            else []
+
+        config = {
+            "hostname": self.vm_hash,
+            "disable_root": False,
+            "ssh_pwauth": True,
```
Comment (hoh) [inline vm_supervisor/vm/firecracker/instance.py:135]: Password authentication should never be allowed outside of tests.

```suggestion
            "ssh_pwauth": False,
```

File: vm_supervisor/vm/firecracker/instance.py [hunk 6]
```diff
@@ -115,5 +120,89 @@ async def wait_for_init(self) -> None:
...
+                        "dhcp4": False,
+                        "dhcp6": False,
+                        "addresses": [
+                            str(ip)
+                        ],
+                        "gateway4": route,
+                        "nameservers": {
+                            "addresses": [
+                                "8.8.8.8",
+                                "8.8.4.4",
```
Comment (hoh) [inline vm_supervisor/vm/firecracker/instance.py:168]: The name servers from the settings should be used.

File: vm_supervisor/vm/firecracker/instance.py [hunk 7]
```diff
@@ -115,5 +120,89 @@ async def wait_for_init(self) -> None:
...
+            yaml_file.write(config_output)
+            yaml_file.flush()
+
+        return yaml_file.name
+
+    async def _create_network_file(self) -> str:
+        """Creates network configuration file for cloud-init tool"""
+
+        ip = self.tap_interface.guest_ip.with_prefixlen
+        route = str(self.tap_interface.host_ip).split("/", 1)[0]
```
Comment (hoh) [inline vm_supervisor/vm/firecracker/instance.py:153]: This should be a property of the tap_interface.

Comment (hoh) [inline vm_supervisor/vm/firecracker/instance.py:153]: This should be its own function

**Comment by hoh** [comment]
I refactored the use of temporary files in a commits to squash later: https://github.com/aleph-im/aleph-vm/pull/333/commits/9c8cbe2f23eeb07dd88620d3f198543c0a1e8cf3

File: kernels/microvm-kernel-x86_64-5.10.config (hunk 1 above)
Comment (nesitor) [reply kernels/microvm-kernel-x86_64-5.10.config:2159]: Yes, I recompiled the kernel using that config and everything works

File: vm_supervisor/vm/firecracker/instance.py [hunk 8]
```diff
@@ -115,5 +122,79 @@ async def wait_for_init(self) -> None:
 
     async def configure(self):
         """Configure the VM by sending configuration info to it's init"""
```
Comment (hoh) [inline vm_supervisor/vm/firecracker/instance.py:124]:
```suggestion
        """Configure the VM by sending configuration info to it's init"""
        # Configuration of instances is sent during `self.setup()` by passing it via a volume.
```

**Review by hoh — APPROVED** [review]

**Pull request merged**